src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -964,112 +1020,227 @@ Transaction::pointer NetworkOPsImp::processTransactionCb (
         // signature not checked
         std::string reason;
 
-        if (! trans->checkSign (reason))
+        if (! transaction->checkSign (reason))
         {
             m_journal.info << "Transaction has bad signature: " << reason;
-            trans->setStatus (INVALID);
-            trans->setResult (temBAD_SIGNATURE);
-            getApp().getHashRouter ().setFlag (trans->getID (), SF_BAD);
-            return trans;
+            transaction->setStatus (INVALID);
+            transaction->setResult (temBAD_SIGNATURE);
+            getApp().getHashRouter ().setFlag (transaction->getID (), SF_BAD);
+            return;
         }
 
-        getApp().getHashRouter ().setFlag (trans->getID (), SF_SIGGOOD);
+        getApp().getHashRouter ().setFlag (transaction->getID (), SF_SIGGOOD);
+
+        if (bLocal)
+            doTransactionSync (transaction, bAdmin, failType);
+        else
+            doTransactionAsync (transaction, bAdmin, failType);
+    }
+}
+
+void NetworkOPsImp::doTransactionAsync (Transaction::pointer transaction,
+        bool bAdmin, FailHard failType)
+{
+    std::lock_guard<std::mutex> lock (mMutex);
+
+    mTransactions.push_back (TransactionStatus (transaction, bAdmin, false,
+        failType));
+
+    if (mDispatchState == DispatchState::none)
+    {
+        m_job_queue.addJob (jtBATCH, "transactionBatch",
+                std::bind (&NetworkOPsImp::transactionBatch, this));
+        mDispatchState = DispatchState::scheduled;
     }
+}
+
+void NetworkOPsImp::doTransactionSync (Transaction::pointer transaction,
+        bool bAdmin, FailHard failType)
+{
+    std::unique_lock<std::mutex> lock (mMutex);
 
+    mTransactions.push_back (TransactionStatus (transaction, bAdmin, true,
+        failType));
+
+    do
     {
-        auto lock = beast::make_lock(getApp().getMasterMutex());
+        if (mDispatchState == DispatchState::running)
+        {
+            // A batch processing job is already running, so wait.
+            mCond.wait (lock);
+        }
+        else
+        {
+            std::vector<TransactionStatus> transactions;
 
-        bool didApply;
-        TER r = m_ledgerMaster.doTransaction (
-            trans->getSTransaction (),
-            bAdmin ? (tapOPEN_LEDGER | tapNO_CHECK_SIGN | tapADMIN)
-            : (tapOPEN_LEDGER | tapNO_CHECK_SIGN), didApply);
-        trans->setResult (r);
+            mTransactions.swap (transactions);
+            mDispatchState = DispatchState::running;
+            lock.unlock();
```

Comment:
It's as Howard says.

And the point of scoped mutexes is for RAII so that it will be unlocked in case of an exception. Unlocking it before the end of scope doesn't interfere with that. And locking it later on within the scope still allows RAII to clean up when scope ends as normal.